### PR TITLE
Dp/loss

### PR DESCRIPTION
**Pull request by dpanici**
- Add ability to pass in a custom loss function to any objective, such as `jnp.mean` to take the average as the target or `jnp.max` to take the max as the target

### CODE REVIEW

File: desc/objectives/_bootstrap.py [hunk 1]
```diff
@@ -49,6 +49,10 @@ class BootstrapRedlConsistency(_Objective):
         Whether target and bounds should be normalized before comparing to computed
         values. If `normalize` is `True` and the target is in physical units,
         this should also be set to True.
+    loss_function : function, optional
```
Comment (f0uriest) [inline desc/objectives/_bootstrap.py:52]: This should be `callable` not `function`

File: desc/objectives/_generic.py [hunk 2]
```diff
@@ -201,6 +207,14 @@ class GenericObjective(_Objective):
         Whether target and bounds should be normalized before comparing to computed
         values. If `normalize` is `True` and the target is in physical units,
         this should also be set to True.
+    loss_function : function, optional
+        User-defined loss function to apply to the objective values once computed.
+        Must be a JAX transformable function, e.g. `jnp.mean` for taking the average
+        or `lambda x: 3*x`, etc
+    loss_function : function, optional
```
Comment (f0uriest) [inline desc/objectives/_generic.py:214]: duplicated

File: desc/objectives/_generic.py [hunk 3]
```diff
@@ -318,6 +334,14 @@ class ToroidalCurrent(_Objective):
...
```
Comment (f0uriest) [inline desc/objectives/_generic.py:341]: duplicated

File: desc/objectives/_generic.py [hunk 4]
```diff
@@ -495,7 +521,11 @@ class RotationalTransform(_Objective):
         values. If `normalize` is `True` and the target is in physical units,
         this should also be set to True.
         Note: has no effect for this objective.
-    grid : Grid, optional
+    loss_function : function, optional
+        User-defined loss function to apply to the objective values once computed.
+        Must be a JAX transformable function, e.g. `jnp.mean` for taking the average
+        or `lambda x: 3*x`, etc
+    grid : Grid, ndarray, optional
```
Comment (f0uriest) [inline desc/objectives/_generic.py:528]: does this accept ndarrays? I thought all the objectives only took `Grid` objects?

File: desc/objectives/_qs.py [hunk 5]
```diff
@@ -37,6 +37,13 @@ class QuasisymmetryBoozer(_Objective):
         Whether target and bounds should be normalized before comparing to computed
         values. If `normalize` is `True` and the target is in physical units,
         this should also be set to True.
+    loss_function : function, optional
+        User-defined loss function to apply to the objective values once computed.
+        Must be a JAX transformable function, i.e. jnp.mean for taking the average
+    loss_function : function, optional
```
Comment (f0uriest) [inline desc/objectives/_qs.py:43]: duplicated

File: desc/objectives/objective_funs.py [hunk 6]
```diff
@@ -742,6 +747,15 @@ def __init__(
         assert normalize_target in {True, False}
         assert (bounds is None) or (isinstance(bounds, tuple) and len(bounds) == 2)
         assert (bounds is None) or (target is None), "Cannot use both bounds and target"
+        assert callable(loss_function), "Loss function must be callable!"
+        assert (
```
Comment (f0uriest) [inline desc/objectives/objective_funs.py:751]: this might be too restrictive, eg
``len(getfullargspec(jnp.max).args)`` is 6, but all the other args have default values. Besides, the next line will check if its valid when called with a single argument.

File: desc/objectives/objective_funs.py [hunk 7]
```diff
@@ -742,6 +747,15 @@ def __init__(
...
+            len(getfullargspec(loss_function).args) == 1
+        ), "Loss function must accept a single array as an argument!"
+        _test_output_loss = loss_function(jnp.array([1, 1]))
```
Comment (f0uriest) [inline desc/objectives/objective_funs.py:754]: might be better to pass an array of size `dim_f` here and then update the `dim_f` attribute afterwards with the correct output size.

**Review by f0uriest — CHANGES_REQUESTED** [review]
Left a few comments to be addressed

One other thing is we should make it clear in the docs somewhere (possibly in the docstring) when the loss function is applied, ie before/after weighting and bounds etc.

File: desc/objectives/objective_funs.py [hunk 8]
```diff
@@ -902,16 +919,18 @@ def compute(self, *args, **kwargs):
 
     def compute_unscaled(self, *args, **kwargs):
         """Compute the raw value of the objective."""
-        return jnp.atleast_1d(self.compute(*args, **kwargs))
+        return jnp.atleast_1d(
+            self._loss_function(jnp.atleast_1d(self.compute(*args, **kwargs)))
```
Comment (ddudt) [inline desc/objectives/objective_funs.py:923]: Does this need the extra `jnp.atleast_1d` call here? `computed_scaled` and `compute_scaled_error` don't have this.

**Review by ddudt — COMMENTED** [review]
This is cool. It would be nice to add something similar for the plotting functions too (not in this PR).

File: desc/objectives/_generic.py (hunk 4 above)
Comment (dpanici) [reply desc/objectives/_generic.py]: it does not, some of the docstrings still had ndarray but I think they all only take Grid so I've updated them accordingly

**Comment by f0uriest** [comment]
Should also update ``docs/adding_objectives.rst`` with the new stuff.

Also, will everything work correctly if the loss function results in the output being different from ``dim_f``?

File: desc/objectives/linear_objectives.py
[the remark this answers is not in the document]
Comment (dpanici) [reply desc/objectives/linear_objectives.py:1]: no I had made some changes and I guess my black did not remove those blank lines, so I had to do it myself

**Comment by dpanici** [comment]
> Should also update `docs/adding_objectives.rst` with the new stuff.
> 
> Also, will everything work correctly if the loss function results in the output being different from `dim_f`?

Not yet, I have not tested it enough.

File: desc/objectives/objective_funs.py (hunk 8 above)
Comment (dpanici) [reply desc/objectives/objective_funs.py:931]: I put it in case loss function returned something that was not a 1D ndarray (like jnp.mean() which returns a 0D array)

**Comment by dpanici** [comment]
Need to think of how to change weights etc, if jnp.mean makes the dimension 1, but weights is `dim_f` based off grid. If fxn is linear/affine, should be ok if we apply after the weights, but other functions (softmin etc) it would matter

think of how broadcasting will work if loss_function changes the dimension

**Comment by dpanici** [comment]
- [ ] add option to say where to apply loss
- [ ] change dim_f according to the dim of the loss fnx

File: desc/objectives/_bootstrap.py [hunk 9]
```diff
@@ -49,7 +49,20 @@ class BootstrapRedlConsistency(_Objective):
...
         values. If `normalize` is `True` and the target is in physical units,
         this should also be set to True.
-    grid : Grid, ndarray, optional
+    loss_function : callable, optional
+        User-defined loss function to apply to the objective values once computed.
+        Must be a JAX transformable function, e.g. `jnp.mean` for taking the average
+        or `lambda x: 3*x`, etc.
+        This loss function is called on the raw compute value, before any shifting,
+        scaling, or normalization.
+    where_apply_loss " str,
```
Comment (f0uriest) [inline desc/objectives/_bootstrap.py:58]: `"` should be `:`

also in a bunch of other places

Comment (f0uriest) [reply desc/objectives/_bootstrap.py:58]: also don't need comma after `str`

File: desc/objectives/_bootstrap.py [hunk 10]
```diff
@@ -49,7 +49,20 @@ class BootstrapRedlConsistency(_Objective):
...
+        where to apply the user defined loss function. One of "compute_unscaled"
```
Comment (f0uriest) [inline desc/objectives/_bootstrap.py:59]: i'd vote to rename these options to just `scaled` and `unscaled`, feels a bit clearer

File: desc/objectives/_generic.py [hunk 11]
```diff
@@ -52,7 +52,27 @@ class ObjectiveFromUser(_Objective):
...
+        This loss function is called on the raw compute value, before any shifting,
+        scaling, or normalization.
+    where_apply_loss " str,
+        where to apply the user defined loss function. One of "compute_unscaled"
+        or "compute_scaled"
+        If "compute_unscaled", the loss function will apply to
+        the raw objective function values, before the target is applied.
+        If "compute_scaled", the loss function will apply to
+        the scaled objective function values, after the target is applied.
+    where_apply_loss " str,
```
Comment (f0uriest) [inline desc/objectives/_generic.py:68]: duplicated text

File: desc/objectives/_generic.py [hunk 12]
```diff
@@ -215,7 +239,27 @@ class GenericObjective(_Objective):
...
```
Comment (f0uriest) [inline desc/objectives/_generic.py:255]: duplicated text

File: desc/objectives/_generic.py [hunk 13]
```diff
@@ -350,7 +398,27 @@ class ToroidalCurrent(_Objective):
...
```
Comment (f0uriest) [inline desc/objectives/_generic.py:414]: duplicated text

File: desc/objectives/objective_funs.py [hunk 14]
```diff
@@ -760,6 +760,20 @@ class _Objective(IOAble, ABC):
...
+        This loss function is called on the raw compute value, before any shifting,
+        scaling, or normalization.
+    where_apply_loss " str,
+        where to apply the user defined loss function. One of "compute_unscaled"
+        or "compute_scaled"
+        If "compute_unscaled", the loss function will apply to
+        the raw objective function values, before the target is applied.
+        If "compute_scaled", the loss function will apply to
+        the scaled objective function values, after the target is applied.
+
```
Comment (f0uriest) [inline desc/objectives/objective_funs.py:776]: extra blank line?

File: desc/objectives/objective_funs.py [hunk 15]
```diff
@@ -786,13 +800,44 @@ def __init__(
...
         assert np.all(np.asarray(weight) > 0)
         assert normalize in {True, False}
         assert normalize_target in {True, False}
         assert (bounds is None) or (isinstance(bounds, tuple) and len(bounds) == 2)
         assert (bounds is None) or (target is None), "Cannot use both bounds and target"
+        assert callable(loss_function), "Loss function must be callable!"
+        # function should be able to handle input of this type and shape
+        try:
+            _test_output_loss = loss_function(jnp.ones((2,)))
```
Comment (f0uriest) [inline desc/objectives/objective_funs.py:829]: any reason the user couldn't create a loss function that specialized on the shape of the expected input? Would this break that?

File: desc/objectives/objective_funs.py [hunk 16]
```diff
@@ -885,6 +933,7 @@ def jit(self):
         except AttributeError:
             pass
         self.compute_scaled = jit(self.compute_scaled)
+        self.compute_scaled2 = self.compute_scaled
```
Comment (f0uriest) [inline desc/objectives/objective_funs.py:936]: is this used?

File: desc/objectives/objective_funs.py (hunk 15 above)
Comment (dpanici) [reply desc/objectives/objective_funs.py:829]: No, this was just to test that the function accepts arrays. There is probably a better way by having it accept dim_f size array and ensuring no error is raised

File: desc/objectives/objective_funs.py (hunk 16 above)
Comment (dpanici) [reply desc/objectives/objective_funs.py:936]: no it was a debug thing, thanks for catching

**Comment by dpanici** [comment]
@f0uriest @ddudt I am unsure how to deal with if someone want to say do mean iota on the unscaled values. That would change the dim_f of the compute_unscaled to be just 1 instead of `grid.num_rho`. However, `compute_scaled` will be still size `num_rho` because we multiply by a  extra weighting in the `_scale` method of `RotationalTransform`. This would cause issues because the dim_f would be different for these two methods, but that means the scaled and unscaled jacobians are different. 

Does it even make sense to apply weighting after a reduction operation like mean is used on the unscaled values?

**Comment by dpanici** [comment]
- Clear up quadrature vs objective weights #484 
 - possibly apply loss to quadrature weights
 - specify that is only correct for mean/min/max
   - str arg
   - if scalar, dont use quadrature weights in compute scaled

**Comment by dpanici** [comment]
Just have 3 predefined, then one called 'callable' and it will be basically a custom compute scaled, and have it accept f, weights, etc, then have user decide what they want to do (where to apply loss etc)

**Review by f0uriest — APPROVED** [review]
